perl5db.pl: "c subname" and "b subname" break on first executable line if possible

tonycoz opened the pull request:
Fixes #799

tonycoz committed:
perl5db.pl: it fiddles with %sub, not @sub

tonycoz committed:
perl5db.pl: b subname and c subname break on first executable line

This currently doesn't try to handle "b postpone subname" since that
internally has an offset function that doesn't really work
with this implementation.

This is a partial fix for #799

mauke commented on lib/perl5db.pl:
```diff
@@ -2033,19 +2033,12 @@ sub _DB__handle_c_command {
-
-            # Force the line number to be numeric.
-            $i = $i + 0;
+            my ($file, $s, $e) = do { subroutine_first_breakable_line($subname) };
```
Did you mean for this `do` to be an `eval`?

mauke commented on lib/perl5db.pl:
```diff
@@ -2058,19 +2051,18 @@ sub _DB__handle_c_command {
             else {
-                print $OUT "Subroutine $subname not found.\n";
-                next CMD;
+                  print $OUT $@;
+                  next CMD;
```
Why the extra indentation here?

mauke commented on lib/perl5db.pl:
```diff
@@ -5389,6 +5381,61 @@ sub subroutine_filename_lines {
     return (find_sub($subname) =~ /^(.*):(\d+)-(\d+)$/);
 } ## end sub subroutine_filename_lines
 
+=head2 subroutine_first_breakable(subname)
```
Did you mean: `subroutine_first_breakable_line`?

mauke commented on lib/perl5db.pl:
```diff
@@ -5389,6 +5381,61 @@ sub subroutine_filename_lines {
 
+=head2 subroutine_first_breakable(subname)
+
+Return the file and line number of the first breakable line in
+subname.
```
It actually returns three values.

mauke commented on lib/perl5db.pl:
```diff
@@ -5401,9 +5448,8 @@ sub break_subroutine {
-    my ( $file, $s, $e ) = subroutine_filename_lines($subname)
-      or die "Subroutine $subname not found.\n";
-
+    my ( $file, $s, $e ) = eval { subroutine_first_breakable_line($subname) }
+      or die $@;
```
`eval { foo() } or die $@` could be simplified to just `foo()`.

tonycoz commented:
What I pushed last week was a bit undercooked.

@mauke, thanks for the feedback, I think I've fixed each of the problems you mentioned, and a few more.

mauke commented on lib/perl5db.pl:
```diff
@@ -5389,6 +5381,60 @@ sub subroutine_filename_lines {
+  my ($file, $sline, $eline) = subroutine_first_breakable_line($subname);
+
+It first tries to find the line number and file of the first OP in
+subname, if that fails it falls back to returning the range of lines
```
I'd punctuate this as `... subname. If that fails, it falls back ...`.

I feel like the second sentence is missing something. Maybe "... the range of all lines covered by subname"? My understanding at least is that we first try to be precise and return the line of the first OP exactly; if that fails, we give up and just return the range of all line numbers that lie within the given sub (and our caller has to deal with it).

tonycoz replied:
Thinking about it, the way the function behaved made it difficult to document (and use), so it's behaviour is now closer to what the name promises, and now leaves less work for callers.

mauke commented on lib/perl5db.pl:
```diff
@@ -2055,22 +2048,13 @@ sub _DB__handle_c_command {
-                }
-                $i = $_line_num;
+                $i = $line;
             } ## end if ($i)
```
Just noticed this: The comment should be either removed or changed to `end if ($line)`.

mauke commented:
Looks good to me.

tonycoz committed:
perl5db.pl: apply the "break on first line" fix to b postpone subname

and just plain fix "b postpone subname", which didn't stop at all
on my postponed sub test code at all as far back at 5.10.

Fixes #799

tonycoz committed:
perl5db.t: add a test for "b compile subname"

I've just been looking at the implementation, and there's no
test, so add one.

tonycoz commented:
added b postpone support
